// eslint-bridge/lib/rules/main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const code_eval_1 = require("./code-eval");
const cookies_1 = require("./cookies");
const cors_1 = require("./cors");
const encryption_1 = require("./encryption");
const file_handling_1 = require("./file-handling");
const hashing_1 = require("./hashing");
const http_endpoint_1 = require("./http-endpoint");
const http_requests_1 = require("./http-requests");
const os_command_1 = require("./os-command");
const process_argv_1 = require("./process-argv");
const pseudo_random_1 = require("./pseudo-random");
const regular_expr_1 = require("./regular-expr");
const sockets_1 = require("./sockets");
const sql_queries_1 = require("./sql-queries");
const standard_input_1 = require("./standard-input");
const xpath_1 = require("./xpath");
const ruleModules = {};
exports.rules = ruleModules;
ruleModules["code-eval"] = code_eval_1.rule;
ruleModules["cookies"] = cookies_1.rule;
ruleModules["cors"] = cors_1.rule;
ruleModules["encryption"] = encryption_1.rule;
ruleModules["file-handling"] = file_handling_1.rule;
ruleModules["hashing"] = hashing_1.rule;
ruleModules["http-endpoint"] = http_endpoint_1.rule;
ruleModules["http-requests"] = http_requests_1.rule;
ruleModules["os-command"] = os_command_1.rule;
ruleModules["process-argv"] = process_argv_1.rule;
ruleModules["pseudo-random"] = pseudo_random_1.rule;
ruleModules["regular-expr"] = regular_expr_1.rule;
ruleModules["sockets"] = sockets_1.rule;
ruleModules["sql-queries"] = sql_queries_1.rule;
ruleModules["standard-input"] = standard_input_1.rule;
ruleModules["xpath"] = xpath_1.rule;
